pkg/tool/osdctl/osdctl.go

```diff
@@ -0,0 +1,241 @@
+	checksumFile, err := os.Open(checksumFilePath)
+	if err != nil {
+		return fmt.Errorf("failed to open osdctl checksum file '%s': %w", checksumFilePath, err)
+	}
+	var checksum string
+	scanner := bufio.NewScanner(checksumFile)
+	for scanner.Scan() {
+		line := scanner.Text()
+		fmt.Printf("'%s'\n", line)
+		tokens := strings.Fields(line)
+		if len(tokens) != 2 {
+			return fmt.Errorf("the checksum file '%s' is invalid", checksumFile.Name())
+		}
+
+		if osdctlBinaryAsset.GetName() != tokens[1] {
+			continue
+		}
+		checksum = tokens[0]
+	}
+	if err = checksumFile.Close(); err != nil {
+		return fmt.Errorf("failed to close checksumfile")
+	}
+
+	if strings.TrimSpace(binarySum) != strings.TrimSpace(checksum) {
+		fmt.Printf("WARNING: Checksum for osdctl does not match the calculated value. Please retry installation. If issue persists, this tool can be downloaded manually at %s\n", osdctlBinaryAsset.GetBrowserDownloadURL())
+		// We shouldn't link this binary to latest if the checksum isn't valid
+		return nil
+	}
+	// Untar osdctl file
+	err = unArchive(osdctlBinaryFilepath, versionedDir)
+	if err != nil {
+		return fmt.Errorf("failed to unarchive the osdctl asset file '%s': %w", filepath.Join(versionedDir, osdctlBinaryAsset.GetName()), err)
+	}
+
+	// Link as latest
+	latestFilePath := t.symlinkPath(latestDir)
+	err = os.Remove(latestFilePath)
+	if err != nil && !os.IsNotExist(err) {
+		return fmt.Errorf("failed to remove existing 'osdctl' binary at '%s': %w", latestDir, err)
+	}
+	err = os.Symlink(filepath.Join(versionedDir, "osdctl"), latestFilePath)
+
+	if err != nil {
+		return fmt.Errorf("failed to link new 'osdctl' binary to '%s': %w", latestDir, err)
+	}
+	return nil
+}
+
+// toolDir returns this tool's specific directory given the root directory all tools are installed in
+func (t *Tool) toolDir(rootDir string) string {
+	return filepath.Join(rootDir, "osdctl")
+}
+
+func (t *Tool) symlinkPath(latestDir string) string {
+	return filepath.Join(latestDir, "osdctl")
+}
+
+func (t *Tool) Remove(rootDir, latestDir string) error {
+	// Remove all binaries owned by this tool
+	toolDir := t.toolDir(rootDir)
+	err := os.RemoveAll(toolDir)
+	if err != nil {
+		return fmt.Errorf("failed to remove %s: %w", toolDir, err)
+	}
+
+	// Remove all symlinks owned by this tool
+	latestFilePath := t.symlinkPath(latestDir)
+	err = os.Remove(latestFilePath)
+	if err != nil {
+		return fmt.Errorf("failed to remove symlinked file %s: %w", latestFilePath, err)
+	}
+	return nil
+}
+
+func unArchive(source string, destination string) error {
+	src, err := os.Open(source)
+	if err != nil {
+		return fmt.Errorf("failed to open tarball '%s': %v", source, err)
+	}
+	defer func() {
+		err = src.Close()
+		if err != nil {
+			fmt.Printf("warning: failed to close '%s': %v\n", src.Name(), err)
+		}
+	}()
+	uncompressed, err := gzip.NewReader(src)
+	if err != nil {
+		return fmt.Errorf("failed to read the new gzip reader '%s': %v", source, err)
```

Comment:
`"failed to read the new gzip reader '%s': %v"` -> `"failed to read the gzip file '%s': %v"`